domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/core/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+)
+
+// State describes retrieval and persistence methods for upgrade info.
+type State interface {
+	// AddUser will add a new user to the database. If the user already exists
+	// an error that satisfies usererrors.AlreadyExists will be returned.
+	AddUser(context.Context, string, string, string, string) error
+
+	// GetUser will retrieve the user specified by name from the database where
+	// the user is active and has not been deleted. If the user does not exist
+	// or is deleted an error that satisfies usererrors.NotFound will be
+	// returned.
+	GetUser(context.Context, string) (user.User, error)
+	// RemoveUser marks the user as deleted. This obviates the ability of a user
+	// to function, but keeps the user retaining provenance, i.e. auditing.
+	RemoveUser(context.Context, string) error
+	// SetPassword removes any active activation keys and sets the user password
```

Comment:
@anvial should we remove the word "active" here? I think all activation keys are active because we remove them once used?

Do you think we should also have two `SetPassword` methods on state?

The first will be `SetPassword` as we currently have it but the second will be `SetPasswordWithKey` or some such where we pass in the activation key for a user and let the database confirm the activation key exists for a user before doing the insert operation?

Also I don't think these functions should be taking a plain text password. They should be taking a string of the hash and a string of the salt. Hasing the plain text away is a responsibility of the services layer.